### ISSUE TITLE
Analyzer implementation of top level type inference doesn't guarantee correct ordering

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#31925]
The analyzer implementation of top level type inference consists of two phases:
- Determine dependencies
- Infer types

Since the types are not available during the "determine dependencies" phase, it is not possible for the analyzer to determine a complete set of dependencies.  For example, consider the following code:

```dart
class C {
  var x = a;
}
var a = ...;
var c = new C();
var d = c.x;
```

The type of `d` depends on the types of `c` and `C.x`, however the dependency on `C.x` can't be determined until the type of `c` has been computed.  As a result, that the analyzer doesn't record a dependency between `d` and `C.x`, so it might or might not assign the correct type to `d`, depending upon the order in which code is visited during analysis.

(Note that the front end solves this problem by performing type inference in a single phase that interleaves both the computation of types and the determination of dependencies, so this is an analyzer-specific problem.)

As part of the work on #31638, we are beginning to add static warnings to the analyzer in order to alert the user to the fact that the analyzed types might be incorrect.  This bug exists to document why the warnings are needed.

**Comment by leafpetersen** [comment]
@stereotype441 Do you have a sense of how likely this is to impact users if we postpone a fix until after Dart 2?  Are there warnings for this in place?

**Comment by stereotype441** [comment]
Yes, to my knowledge all the necessary warnings have been added to the analyzer.

**Comment by leafpetersen** [comment]
Ok, moving this off of stable, but leaving it open to track the missing top level inference work.

**Comment by srawlins** [comment]
We've recently (2.19? 2.18?) changed type inference w.r.t. ordering and dependencies. I'll close this as stale. Please re-open if it is still a valid issue.
